src/modules/commandrunner/src/lib/CommandRunner.cpp

```diff
@@ -1,465 +1,693 @@
-    if (instance.IsCanceled(commandId))
-    {
-        return status;
-    }
-
-    OsConfigLogInfo(CommandRunnerLog::Get(), "Running command '%s'", commandId.c_str());
-    instance.SetCommandIdToRefresh(commandId);
-    instance.UpdatePartialCommandStatus(commandId, 0, initialState);
-
-    if ((nullptr != instance.cacheFunction) && (0 != (status = instance.cacheFunction())))
-    {
-        OsConfigLogError(CommandRunnerLog::Get(), "Unable to persist to cache, skipping command '%s' with %d", commandId.c_str(), status);
-        return status;
-    }
+    bool exists = false;
+    std::lock_guard<std::mutex> lock(m_cacheMutex);
 
-    unsigned int maxTextResultBytes = 0;
-    unsigned int maxPayloadSizeInBytes = static_cast<unsigned int>(instance.GetMaxPayloadSizeInBytes());
-    if (maxPayloadSizeInBytes > 0)
+    if (m_commandMap.find(id) != m_commandMap.end())
     {
-        unsigned int estimatedSize = strlen(commandStatusTemplate) + strlen(commandId.c_str()) + (3 * strlen(resultCodeChar));
-        maxTextResultBytes = (maxPayloadSizeInBytes > estimatedSize) ? (maxPayloadSizeInBytes - estimatedSize) : 1;
+        exists = !m_commandMap[id].expired();
     }
 
-    void* context = nullptr;
-    CommandStatus* commandStatus = nullptr;
-    std::string tmpFilePath;
-    if (nullptr != (commandStatus = instance.GetCommandStatus(commandId)))
-    {
-        tmpFilePath = CommandRunner::GetTmpFilePath(commandStatus->GetUniqueId());
-        context = reinterpret_cast<void*>(const_cast<char*>(tmpFilePath.c_str()));
-    }
+    return exists;
+}
 
-    status = ExecuteCommand(context, command.c_str(), replaceEol, true, maxTextResultBytes, timeoutSeconds, &textResult, &(CommandRunner::CommandExecutionCallback), CommandRunnerLog::Get());
+int CommandRunner::ScheduleCommand(std::shared_ptr<Command> command)
+{
+    int status = 0;
 
-    if (IsFullLoggingEnabled())
+    if (!CommandExists(command->GetId()))
     {
-        OsConfigLogInfo(CommandRunnerLog::Get(), "Command '%s' ('%s') completed with %d and '%s'", commandId.c_str(), command.c_str(), status, textResult ? textResult : "no text result");
+        if (0 == (status = PersistCommandStatus(command->GetStatus())))
+        {
+            if (0 == (status = CacheCommand(command)))
+            {
+                m_commandQueue.Push(command);
+            }
+            else
+            {
+                OsConfigLogError(CommandRunnerLog::Get(), "Failed to cache command with id '%s'", command->GetId().c_str());
+            }
+        }
+        else
+        {
+            OsConfigLogError(CommandRunnerLog::Get(), "Failed to persist command to disk. Skipping command with id '%s'", command->GetId().c_str());
+        }
     }
     else
     {
-        OsConfigLogInfo(CommandRunnerLog::Get(), "Command '%s' completed with %d", commandId.c_str(), status);
+        OsConfigLogError(CommandRunnerLog::Get(), "Command with id '%s' already exists", command->GetId().c_str());

```

Comment:
>Command with id '%s' [](http://example.com/codeflow?start=51&length=20)

Everywhere. Instead of "Command with id '%s'" can instead log "Command '%s'". Shorter lines, smaller binary.